### PR TITLE
Add ParseBuffer::advance_to, the opposite of fork

### PR DESCRIPTION
**Pull request by CAD97**
Addresses (some of?) #479 by allowing speculative parsing without leaving the ParseStream land.

The method in the first commit is unsafe, as it requires the user to guarantee that the fork is indeed forked from the parse stream it is being joined to. This could potentially be guaranteed by examining the state of the two parse streams' Spans, but doing so is not stable.

This is made to panic safely by making self.scope a Rc and using Rc::ptr_eq to make sure that the base and the fork are both using the same root Rc'd scope. The second commit implements this approach.

### CODE REVIEW

**Comment by CAD97** [comment]
Side note: testing with `cargo test --release --all-features` locally seems to be failing to compile:

<details>

```
D:\usr\Documents\Code\Rust\syn>rustup show
Default host: x86_64-pc-windows-gnu

installed toolchains
--------------------

beta-x86_64-pc-windows-gnu
nightly-x86_64-pc-windows-gnu

active toolchain
----------------

nightly-x86_64-pc-windows-gnu (directory override for '\\?\D:\usr\Documents\Code\Rust\syn')
rustc 1.33.0-nightly (daa53a52a 2019-01-17)


D:\usr\Documents\Code\Rust\syn>cargo test --release --all-features
   Compiling syn v0.15.26 (D:\usr\Documents\Code\Rust\syn)
error[E0277]: the trait bound `common::syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:143:19
    |
143 |                 $(SpanlessEq::eq($field, $other))&&*
    |                   ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `common::syntax::tokenstream::ThinTokenStream`
...
292 | spanless_eq_struct!(Mac_; path delim tts);
    | ------------------------------------------ in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `common::syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:143:19
    |
143 |                 $(SpanlessEq::eq($field, $other))&&*
    |                   ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `common::syntax::tokenstream::ThinTokenStream`
...
293 | spanless_eq_struct!(MacroDef; tokens legacy);
    | --------------------------------------------- in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `common::syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:200:39
    |
200 |                             true $(&& SpanlessEq::eq($this, $other))*
    |                                       ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `common::syntax::tokenstream::ThinTokenStream`
...
346 | spanless_eq_enum!(TokenTree; Token(0 1) Delimited(0 1 2));
    | ---------------------------------------------------------- in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:143:19
    |
143 |                 $(SpanlessEq::eq($field, $other))&&*
    |                   ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `syntax::tokenstream::ThinTokenStream`
...
292 | spanless_eq_struct!(Mac_; path delim tts);
    | ------------------------------------------ in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:143:19
    |
143 |                 $(SpanlessEq::eq($field, $other))&&*
    |                   ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `syntax::tokenstream::ThinTokenStream`
...
293 | spanless_eq_struct!(MacroDef; tokens legacy);
    | --------------------------------------------- in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `common::syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:143:19
    |
143 |                 $(SpanlessEq::eq($field, $other))&&*
    |                   ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `common::syntax::tokenstream::ThinTokenStream`
...
292 | spanless_eq_struct!(Mac_; path delim tts);
    | ------------------------------------------ in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `common::syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:143:19
    |
143 |                 $(SpanlessEq::eq($field, $other))&&*
    |                   ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `common::syntax::tokenstream::ThinTokenStream`
...
293 | spanless_eq_struct!(MacroDef; tokens legacy);
    | --------------------------------------------- in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:200:39
    |
200 |                             true $(&& SpanlessEq::eq($this, $other))*
    |                                       ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `syntax::tokenstream::ThinTokenStream`
...
346 | spanless_eq_enum!(TokenTree; Token(0 1) Delimited(0 1 2));
    | ---------------------------------------------------------- in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `common::syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:200:39
    |
200 |                             true $(&& SpanlessEq::eq($this, $other))*
    |                                       ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `common::syntax::tokenstream::ThinTokenStream`
...
346 | spanless_eq_enum!(TokenTree; Token(0 1) Delimited(0 1 2));
    | ---------------------------------------------------------- in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `syn`.
warning: build failed, waiting for other jobs to finish...
error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.
error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `syn`.
warning: build failed, waiting for other jobs to finish...
error: Could not compile `syn`.
warning: build failed, waiting for other jobs to finish...
error[E0277]: the trait bound `syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:143:19
    |
143 |                 $(SpanlessEq::eq($field, $other))&&*
    |                   ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `syntax::tokenstream::ThinTokenStream`
...
292 | spanless_eq_struct!(Mac_; path delim tts);
    | ------------------------------------------ in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:143:19
    |
143 |                 $(SpanlessEq::eq($field, $other))&&*
    |                   ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `syntax::tokenstream::ThinTokenStream`
...
293 | spanless_eq_struct!(MacroDef; tokens legacy);
    | --------------------------------------------- in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `syntax::tokenstream::ThinTokenStream: common::eq::SpanlessEq` is not satisfied
   --> tests\common\eq.rs:200:39
    |
200 |                             true $(&& SpanlessEq::eq($this, $other))*
    |                                       ^^^^^^^^^^^^^^ the trait `common::eq::SpanlessEq` is not implemented for `syntax::tokenstream::ThinTokenStream`
...
346 | spanless_eq_enum!(TokenTree; Token(0 1) Delimited(0 1 2));
    | ---------------------------------------------------------- in this macro invocation
    |
note: required by `common::eq::SpanlessEq::eq`
   --> tests\common\eq.rs:36:5
    |
36  |     fn eq(&self, other: &Self) -> bool;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `syn`.

To learn more, run the command again with --verbose.

**Comment by dtolnay** [comment]
Your nightly compiler may be out of date. Our tests compare Syn against libsyntax's parser which changes often.

**Comment by CAD97** [comment]
With the `Rc::ptr_eq` equivalent for 1.15, this should be fully good now.

**Comment by CAD97** [comment]
New example is removing the turbofish, an example close to all our hearts. (All the other small, reasonable examples of `A* B*` I could come up with have `A` as a subset of `B`, which can just be parsed as `B*` and removing the leading `A`.) I'll add the more in-depth drawbacks later this week.

**Comment by dtolnay** [comment]
Nice, that's a great example.

File: src/parse/discouraged.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    ///         if input.peek(Token![::]) && input.peek3(Token![<]) {
+    ///             return Ok(PathSegment {
+    ///                 ident: ident,
+    ///                 arguments: PathArguments::AngleBracketed(input.parse()?),
+    ///             });
+    ///         }
+    ///         if input.peek(Token![<]) && !input.peek(Token![<=]) {
+    ///             let fork = input.fork();
+    ///             if let Ok(arguments) = fork.parse() {
+    ///                 return Ok(PathSegment {
```
Comment (dtolnay) [inline src/parse/discouraged.rs]: Presumably this should call `advance_to` somewhere.

File: src/parse/discouraged.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    ///
+    /// Also note that if `A` is a subset of `B`, `A* B*` can be parsed by parsing
+    /// `B*` and removing the leading members of `A` from the repetition, bypassing
+    /// the need to involve the downsides associated with speculative parsing.
+    ///
+    /// [`ParseStream::fork`]: ../struct.ParseBuffer.html#method.fork
+    ///
+    /// # Example
+    ///
+    /// Let's [kill the turbofish][RFC#2544]! Specifically, according to [RFC#2544],
```
Comment (dtolnay) [inline src/parse/discouraged.rs]: > Let's kill the turbofish!

I would prefer if my crates' docs couldn't be misinterpreted as taking a position on turbofish. :eyes:

Maybe something milder like:

> There has been chatter about the possibility of making the colons in turbofish syntax like `path::to::<T>` no longer required by accepting `path::to<T>` in expression position.

**Comment by CAD97** [comment]
Rebased on top of master for nightly update and Windows test fixes. Adding the dedicated tests now.

**Comment by dtolnay** [comment]
Please write the test without a transmute_copy. If you parse `"()()"` and advance a stream from inside one set of parens to inside the other set, does that hit the same check?

```
( )   ( )
 ^a    ^b

a.advance_to(&b);
```

Then I think the last thing is a Drawbacks section in the documentation.

**Comment by dtolnay** [comment]
I pushed a fix for the [CI failure](https://travis-ci.org/dtolnay/syn/jobs/483738646).

**Comment by dtolnay** [comment]
Ping on the remaining work. Is this still happening?

- Test the new functionality without a transmute_copy;
- Drawbacks section in the documentation.

**Comment by CAD97** [comment]
I've just not had the time to set up an actual proc-macro usage to check the actual emitted errors. (I'll likely borrow the example in the ex-turbofish poc pr.) This past couple of weeks have been hectic for me.

I'm planning to try to get to it this weekend.

**Comment by dtolnay** [comment]
Friendly poke to keep this on your radar for when you have time.

**Comment by dtolnay** [comment]
Please open a new PR if you get back to working on this. Thanks!

**Pull request closed without merging**